stdenv/check-meta: add note for Flake usage

MatthewCroughan opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Flake users that use a command like `nix build nixpkgs#hello` on a
broken/insecure package will not be able to use an environment variable
to override that behavior, unless they pass `--impure` to the command.
This commit adds a conditional string that appears only if they are in
pure evaluation mode, thereby signalling that they are using Flake
commands.

Here is an example of how the message looks:

```
user: matthew swordfish in nixpkgs on  add-flake-note [$] 
❯ nix build .#linuxPackages_latest-libre.rtl88xxau-aircrack
error: Package ‘rtl88xxau-aircrack-5.15.1-3a6433’ in /nix/store/abygh3wk0w039hg29x9y0ikdv9r11g1h-source/pkgs/os-specific/linux/rtl88xxau-aircrack/default.nix:46 is marked as broken, refusing to evaluate.

       a) To temporarily allow broken packages, you can use an environment variable
          for a single invocation of the nix tools.

            $ export NIXPKGS_ALLOW_BROKEN=1

          Note: For `nix shell`, `nix build`, `nix develop` or any other Nix 2.4+
          (Flake) command, `--impure` must be passed in order to read this
          environment variable.

       b) For `nixos-rebuild` you can set
         { nixpkgs.config.allowBroken = true; }
       in configuration.nix to override this.

       c) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
         { allowBroken = true; }
       to ~/.config/nixpkgs/config.nix.
(use '--show-trace' to show detailed location information)
```

Co-authored-by: pkharvey <kayharvey@protonmail.com>

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

MatthewCroughan committed:
stdenv/check-meta: add note for Flake usage

Flake users that use a command like `nix build nixpkgs#hello` on a
broken/insecure package will not be able to use an environment variable
to override that behavior, unless they pass `--impure` to the command.

Co-authored-by: pkharvey <kayharvey@protonmail.com>

tomberek reviewed (commented):
The note helps clarify and avoid some frustration. Perhaps also a section 4 to showcase something like: https://discourse.nixos.org/t/flakes-with-unfree-licenses/9405/3

SuperSandro2000 reviewed (commented):
Is there a better way to detect if we are under flakes?

MatthewCroughan commented:
> Is there a better way to detect if we are under flakes?

This is how `impure.nix` does it. So I think what I'm doing is fine. I am simply checking to see if `builtins` has `currentSystem` or not. If not, then we must be in pure eval mode. Though it would be awesome if there was a library function that handled this check instead, maybe we could write one.

```
error: attribute 'currentSystem' missing

       at /nix/store/2ighdnh07y10my6466cm6rxv629dgyr9-nixpkgs-src/pkgs/top-level/impure.nix:18:43:

           17|   # (build, in GNU Autotools parlance) platform.
           18|   localSystem ? { system = args.system or builtins.currentSystem; }
             |                                           ^
           19|
(use '--show-trace' to show detailed location information)
```

thiagokokada commented on pkgs/stdenv/generic/check-meta.nix:
```diff
@@ -120,12 +120,19 @@ let
 
            $ export ${remediation_env_var allow_attr}=1
-
+           ${lib.optionalString (! builtins ? currentSystem) flakeNote}
```
I don't see an issue of using this "hack", but maybe it would be better to add a comment saying that maybe we should substitute this later for `isFlake` function.

MatthewCroughan replied:
How do you add a comment into a multiline string? I would be happy to do that if you could figure this out.

thiagokokada replied:
Maybe we could convert this to a function inside a let and put the comment inside the function.

thiagokokada replied:
Anyway, almost sure this works too:

```
${# this is a comment
  lib.optionalString (! builtins ? currentSystem) flakeNote}
```

MatthewCroughan replied:
I have tidied the code up in the latest force push in a way that should satisfy you. I have now put the optional string into the `let` block for `flakeNote`, which is also where I have added the comment. Now all I do to print it is `${flakeNote}` which keeps the formatting as sensible as it can be.

thiagokokada reviewed (commented):
This is a good change BTW.

SuperSandro2000 commented on pkgs/stdenv/generic/check-meta.nix:
```diff
@@ -136,6 +144,14 @@ let
     '';
 
   remediate_insecure = attrs:
+    let
```
One last little note: can we move this one or more levels up to dedupe it?

Other than that LGTM

MatthewCroughan replied:
Done.

Ma27 commented on lib/trivial.nix:
```diff
@@ -205,6 +205,13 @@ rec {
+
+     Type: inPureEvalMode :: bool
+  */
+  inPureEvalMode = ! builtins ? currentSystem;
```
Is it guaranteed that this check will be correct from now on? For context of what I'm talking about, please read the discussion in https://github.com/NixOS/nix/pull/5301 (tl;dr one could assume previously that `builtins ? getFlake` is sufficient to check if the experimental feature is enabled, now it throws an eval-error, but *exists*. So I'd be careful with these kinds of assumptions in general).

also cc @edolstra whom I discussed this with back then.

MatthewCroughan replied:
It's my view that `builtins.inFlake` or `builtins.inPureEvalMode` should exist and be a primop, having looked at some of the Nix code. I would not know how to implement it, but it's the only way to guarantee that these things will continue working.

Ma27 replied:
Perhaps I'm not realizing that we're thinking the same, but let me put it in a slightly different way to be sure: `inFlake`/`inPureEvalMode` should be a primop *if* it exists.

I mean, I'm not sure if that's desirable (@edolstra raised concerns about these hacks in the PR I linked above already), so this is something to be discussed in `NixOS/nix` IMHO.

IIRC you have a global variable called `evalSettings` in `libexpr` and can access it via `evalSettings.pureEval`.

Regarding this PR: I'm :-1: on adding `inPureEvalMode`, instead we should just add the flake-specific thing to the error-message in any case for two reasons:
* We're doing it already for several other cases without checking which actually applies
* People may learn about `--impure` "on the fly" even if they don't use flakes yet. What I mean is, I think a list of all possible fixes is actually a good thing.

MatthewCroughan replied:
I'm happy to do that too, I'll wait for everyone else's review, since this could go either way.

thiagokokada replied:
> instead we should just add the flake-specific thing to the error-message in any case

I kinda prefer this option too, it is more consistent.

However the only issue I can see is that maybe this will promote Flakes and AFAIK our approach is to not divulge it yet since it is experimental. But maybe the ship sailed already (I think many `nixos-*` already has a `--flake` option that show regardless?).

Ma27 replied:
> However the only issue I can see is that maybe this will promote Flakes and AFAIK our approach is to not divulge it yet since it is experimental. But maybe the ship sailed already (I think many nixos-* already has a --flake option that show regardless Flakes is enabled or not?).

Hydra has flakes support, nixos-rebuild & other stuff as well, so this shouldn't be a big deal.

thiagokokada replied:
Yeah, like I said, the ship probably shipped already.

For me, if @MatthewCroughan is fine in removing the if and just always showing the message, I would approve.

Infinisil commented on lib/trivial.nix:
```diff
@@ -205,6 +205,13 @@ rec {
 
+  /* Determine whether the function is being called from inside pure-eval mode
+     by seeing whether `builtins` contains `currentSystem`. If not, we must be in
+     pure-eval mode. This also implies we are in a Flake.
```
Pure evaluation mode doesn't imply flakes, though that's arguably not that common:
```
$ nix-instantiate --eval --pure-eval -E 'builtins ? currentSystem'
false
```

MatthewCroughan commented:
@Infinisil @thiagokokada @Ma27 I believe I have addressed your feedback. I no longer imply that Flakes have anything to do with pure-eval mode in the commits, and the flake note is displayed always, without a condition. I still think that introducing `lib.inPureEvalMode` could be useful to someone in the same way that `lib.inNixShell` is useful.

Ericson2314 commented:
I don't think we should conflate pure eval and flakes. I certainly want to be able to use pure eval without flakes.

MatthewCroughan commented:
> I don't think we should conflate pure eval and flakes. I certainly want to be able to use pure eval without flakes.

There should be nothing left in the latest commits that conflates flakes with pure eval. Have I missed something?

thiagokokada commented on lib/trivial.nix:
```diff
@@ -205,6 +205,13 @@ rec {
+
+     Type: inPureEvalMode :: bool
+  */
+  inPureEvalMode = ! builtins ? currentSystem;
```
While I still think this function to be useful, maybe it is better to introduce this in another PR since it is not used by anything right now :thinking: ?

MatthewCroughan replied:
I've done that here. https://github.com/NixOS/nixpkgs/pull/157480, I have also rebased this PR to remove this commit, as well as removing all of the conflation in the initial commit.

thiagokokada reviewed (approved):
I added a note, pure I am approving here.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/announcing-nixpkgs-unfree/17505/3
